src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -1,79 +1,217 @@
-      val clusterConfig = new ClusterConfig().setGceClusterConfig(gce).setInitializationActions(initActions.asJava)
-      val cluster = new Cluster().setClusterName(clusterName).setConfig(clusterConfig)
-      val request = dataproc.projects().regions().clusters().create(googleProject, dataprocConfig.dataprocDefaultZone, cluster)
-      try {
-        executeGoogleRequest(request)
-      } catch {
-        case e: GoogleJsonResponseException => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Google Request Failed: ${e.getDetails.getMessage}"))
-      }
+      executeGoogleRequest(request) // returns a Firewall
+    } recoverWith {
+      case e: GoogleJsonResponseException =>
+        if (e.getStatusCode == 404)
+          throw FirewallRuleNotFoundException(googleProject, firewallRuleName)
+        throw CallToGoogleApiFailedException(googleProject, firewallRuleName, e.getStatusCode, e.getDetails.getMessage)
+    }
+  }
+
+
+  /* Adds a firewall rule in the given google project. This firewall rule allows ingress traffic through a specified port for all
+     VMs with the network tag "leonardo". This rule should only be added once per project.
+    To think about: do we want to remove this rule if a google project no longer has any clusters? */
+  private def addFirewallRule(googleProject: GoogleProject): Future[ComputeOperation] = {
+    // Create an Allowed object that specifies the port and protocol of rule
+    //val allowedList = firewallRule.allowed.map(allowedPair => new GoogleAllowed().setIPProtocol(allowedPair.protocol).setPorts(allowedPair.port.asJava))
+    val allowed = new Allowed().setIPProtocol(proxyConfig.jupyterProtocol).setPorts(List(proxyConfig.jupyterPort.toString).asJava)
+    // Create a firewall rule that takes an Allowed object, a name, and the network tag a cluster must have to be accessed through the rule
+    val googleFirewallRule = new Firewall()
+      .setName(dataprocConfig.clusterFirewallRuleName)
+      .setTargetTags(List(dataprocConfig.clusterNetworkTag).asJava)
+      .setAllowed(List(allowed).asJava)
+
+    val request = compute.firewalls().insert(googleProject, googleFirewallRule)
+    Future {
+      executeGoogleRequest(request) // returns a ComputeOperation
+    } recoverWith {
+      case e: GoogleJsonResponseException =>
+        throw CallToGoogleApiFailedException(googleProject, dataprocConfig.clusterFirewallRuleName, e.getStatusCode, e.getDetails.getMessage)
+    }
+  }
+
+  /* Create a bucket in the given google project for the initialization files when creating a cluster */
+  def createBucket(googleProject: GoogleProject, bucketName: String): Future[Unit] = {
+    // Create lifecycle rule for the bucket that will delete the bucket after 1 day - for now, init buckets will be
+    // deleted this way, but eventually we will likely want to delete them after the cluster has been initialized
```

Comment:
Interesting - I could try to incorporate this into the cluster monitor poller (maybe as a separate PR to the one I have open).